src/phoenix/trace/llama_index/callback.py

```diff
@@ -381,3 +411,43 @@ def _get_response_output(response: Any) -> Iterator[Tuple[str, Any]]:
     else:
         yield OUTPUT_VALUE, str(response)
         yield OUTPUT_MIME_TYPE, MimeType.TEXT
+
+
+def _get_end_time(event_data: CBEventData, span_events: List[SpanEvent]) -> Optional[datetime]:
+    """
+    A best-effort attempt to get the end time of an event.
+
+    LlamaIndex's callback system does not guarantee that the on_event_end hook is always called, for
+    example, when an error occurs mid-event.
+    """
+    if end_event := event_data.get("end_event"):
+        tz_naive_end_time = _timestamp_to_tz_naive_datetime(end_event.time)
+    elif span_events:
+        last_span_event = span_events[-1]
+        tz_naive_end_time = last_span_event.timestamp
+    else:
+        return None
+    return _tz_naive_to_tz_aware_datetime(tz_naive_end_time)
+
+
+def _timestamp_to_tz_aware_datetime(timestamp: str) -> datetime:
+    """Converts a timestamp string to a timezone-aware datetime."""
+    return _tz_naive_to_tz_aware_datetime(_timestamp_to_tz_naive_datetime(timestamp))
+
+
+def _timestamp_to_tz_naive_datetime(timestamp: str) -> datetime:
+    """Converts a timestamp string to a timezone-naive datetime."""
+    return datetime.strptime(timestamp, TIMESTAMP_FORMAT)
+
+
+def _tz_naive_to_tz_aware_datetime(timestamp: datetime) -> datetime:
+    """Converts a timezone-naive datetime to a timezone-aware datetime."""
+    return timestamp.replace(tzinfo=_LOCAL_TZINFO)
+
+
+def _get_stacktrace(exception: BaseException) -> str:
+    """Gets stacktrace from exception."""
+    exception_type = type(exception)
+    exception_traceback = exception.__traceback__
+    stack_trace_lines = format_exception(exception_type, exception, exception_traceback)
+    return "".join(stack_trace_lines)
```

Comment:
needs a delimiter
```suggestion
    return "\n".join(stack_trace_lines)
```

Reply:
addressed in [4dd7df3](https://github.com/Arize-ai/phoenix/pull/1449/commits/4dd7df31606bcffbc8c7c42456646624351b44c2)